numba/array.py

```diff
@@ -0,0 +1,177 @@
+        """Construct an Array in the memory space defined by 'where'."""
+
+        assert where in ('host', 'gpu')
+        # we need either a prototype or proper type info
+        assert obj is not None or (shape and dtype)
+        self._host = self._gpu = None
+        self._host_valid = self._gpu_valid = False
+        self._allocate(where, obj, dtype, shape, _o2s(dtype, shape, order), copy)
+        if where == 'host':
+            self._host_valid = True
+            t = self._host
+        else:
+            self._gpu_valid = True
+            t = self._gpu
+        self.shape = t.shape
+        self.strides = t.strides
+        self.dtype = t.dtype
+        self.ndim = t.ndim
+        self.size = t.size
+
+    def host(self):
+        """Get a host array."""
+
+        self._sync('host')
+        #self._invalidate('gpu') # only if write-access is requested !
+        return self._host
+
+    def host_changed(self):
+        """Mark the host array as changed, broadcast updates if needed."""
+
+        self._invalidate('gpu')
+        # only sync if there are active views
+        if self._gpu is not None and sys.getrefcount(self._gpu) > 2:
+            self._sync('gpu')
+
+    def gpu(self):
+        """Get a GPU array."""
+
+        self._sync('gpu')
+        #self._invalidate('host') # only if write-access is requested !
+        return self._gpu
+
+    def gpu_changed(self):
+        """Mark the gpu array as changed, broadcast updates if needed."""
+
+        self._invalidate('host')
+        # only sync if there are active views
+        if self._host is not None and sys.getrefcount(self._host) > 2:
+            self._sync('host')
+
+    def __array__(self, *args):
+
+        self._sync('host')
+        return numpy.array(self._host, *args)
+
+    def _sync(self, where):
+        """Sync the data in one memory space with the other."""
+
+        if where == 'gpu':
+            if self._gpu is None:
+                self._allocate('gpu', None, self.dtype, self.shape, self.strides)
+            if not self._gpu_valid:
+                self._copy_to_gpu()
+        else:
+            if self._host is None:
+                self._allocate('host', None, self.dtype, self.shape, self.strides)
+            if not self._host_valid:
+                self._copy_to_host()
+
+    @trace
+    def _invalidate(self, where):
+        """Mark the host / device array as out-of-date."""
+
+        if where == 'gpu':
+            self._gpu_valid = False
+        else:
+            self._host_valid = False
+
+    @trace
+    def _allocate(self, where, obj=None, dtype=None, shape=None, strides=None,
+                  copy=True):
+        if dtype:
+            dtype = numpy.dtype(dtype)
+        if where == 'host':
+            if obj is not None:
+                self._host = numpy.array(obj, dtype, copy=copy)
+            else:
+                self._host = numpy.empty(shape, dtype, _s2o(dtype, shape, strides))
+        else:
```

Comment:
Yes, definitely. But to keep things simple, I wanted to focus on the 'host'/'cuda' split first. There are multiple improvements to consider (but again, only once this current work is stable and merged):

* support other GPU backends than CUDA, such as HSA, or even OpenCL.
* support multiple GPUs (each with their own memory space, and therefore device array representation)
* support for other non-GPU devices / memory spaces

In that sense I see the current work only as a prototype, and for clarity, would like to avoid being too general (well, other than having the 'where' flag be a string :-) )